### ISSUE TITLE
Unit test failure of the TestAsynchronousExecutionPolicy

### ISSUE THREAD

**Issue by vkarak** [eth-cscs/reframe#78]
The failure happens when running the unittest suite on a local machine (laptop) with 2 Codes - 4 Threads. The failure is the following:
```
FAIL: test_kbd_interrupt_in_wait_with_limited_concurrency (unittests.test_policies.TestAsynchronousExecutionPolicy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/manitart/reframe_dev_local/unittests/test_policies.py", line 310, in test_kbd_interrupt_in_wait_with_limited_concurrency
    self._run_checks(checks, 2)
  File "/home/manitart/reframe_dev_local/unittests/test_policies.py", line 289, in _run_checks
    self.assertEqual(4, self.runner.stats.num_cases())
AssertionError: 4 != 3
```

Internal issue: https://madra.cscs.ch/scs/reframe/issues/582

**Comment by vkarak** [comment]
This happens sporadically and appeared also once during running by Travis CI.

**Comment by vkarak** [comment]
This has become recurring. We should address it asap.

**Comment by vkarak** [comment]
I spotted the problem and it appears to be caused by the (wrong) assumption the unit tests make that none of the submitted tasks would finish at the time when the partition task limit is reached. When this limit is reached, all the currently running tasks are polled. If any of them has finished (in this case the `KeyboardInterruptCheck`), it will be waited for, and this is where the `KeyboardInterrupt` is raised. At this point not all of the tasks may have been submitted leading to this unit test failure. Here is a bit of code to always trigger the failure:

```diff
diff --git a/reframe/frontend/executors/policies.py b/reframe/frontend/executors/policies.py
index 3f4d4a7..e8bc481 100644
--- a/reframe/frontend/executors/policies.py
+++ b/reframe/frontend/executors/policies.py
@@ -189,6 +189,9 @@ class AsynchronousExecutionPolicy(ExecutionPolicy, TaskEventListener):

             self._failall(e)
             raise
+        finally:
+            import time
+            time.sleep(.5)

     def _poll_tasks(self):
         """Update the counts of running checks per partition."""
```

**Comment by vkarak** [comment]
This seems not to be fixed entirely. It still happens from time to time on Travis CI. Perhaps we need to increase the sleep time of the `KeyboardInterruptCheck`.

**Comment by vkarak** [comment]
The problem currently is a bit different. The failure we get is the following:

```
======================================================================
FAIL: test_kbd_interrupt_in_wait_with_concurrency (unittests.test_policies.TestAsynchronousExecutionPolicy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/eth-cscs/reframe/unittests/test_policies.py", line 299, in test_kbd_interrupt_in_wait_with_concurrency
    self._run_checks(checks, 4)
  File "/home/travis/build/eth-cscs/reframe/unittests/test_policies.py", line 289, in _run_checks
    self.assertEqual(4, self.runner.stats.num_failures())
AssertionError: 4 != 3
```

**Comment by vkarak** [comment]
It seems that at the time the `KeyboardInterruptCheck` fails, one of the `SleepCheck`s has already finished. This is likely to happen, because all of the tests have now the same runtime more or less (1s). I think that increasing the sleep time of the `SleepCheck`s in this test to 1.5s-2s will allow it to pass safely.

The general idea for this test is to allow enough time for all the four checks to be submitted and at the same time we need the `KeyboardInterruptCheck` to finish first, so as to make the framework kill the remaining three.

**Comment by vkarak** [comment]
Fixed by #154.
